include/meta/meta.hpp

```diff
@@ -934,13 +942,31 @@ namespace meta
     template <typename... Args>
     using transform = eval<detail::transform_<Args...>>;
 
+    ////////////////////////////////////////////////////////////////////////////
+    // filter
+    /// \cond
+    template <typename Predicate> struct filter_
+    {
+        template <typename State, typename A>
+        using apply = meta::if_<meta::apply<Predicate, A>,
+                                meta::push_back<State, A>, State>;
+    };
+    /// \endcond
+
+    /// Returns a new meta::list where all the elements of \p List A that
+    /// satisfy the Metafunction Class \p Predicate such that
+    /// `apply<Pred,A>::%value` is \c true have been removed.
+    template <typename List, typename Predicate>
+    using filter = meta::foldl<List, meta::list<>, filter_<Predicate>>;
```

Comment:
> I took a look at the MPL lambda facility but it is a beast.

It is a beast. And it has weird corner cases where it doesn't work as the user expects(why are there specializations for `if_`?). 

However, you could also look at tick library [here](https://github.com/pfultz2/Tick/blob/master/tick/detail/matches.h). The `replace_args` is what would be needed to implement `lambda`. So if you write `replace_args<some_metafunction<_1, _2>, a, b>::type` it will be `some_metafunction<a, b>`. It also supports unamed placeholders(ie `_`) but only for simple cases.

Now in tick I cheat and use `std::tuple` to get the ith argument. Because there could be overhead in this approach, I specialize it for the most common use case. Ideally in meta it would be better to not do it this way.

Now [here](https://github.com/pfultz2/Tick/blob/master/tick/detail/matches.h) the `replace_args_recursive` does specializations for `quote` and such. In meta, you would have specializations instead for `lazy_quote`, `lazy_eval`, and `lambda`.

Also, I think that `lambda` should always be explicit instead of implicit inside of `apply`(ie you always write `lambda<some-placeholder-exression>` to create a metafunction class). It just leads to too many corner cases where things break.